docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+process templates.*
+
+* Parameter definitions include metadata which describes the purpose of the parameter.  Since parameter definitions are part of the template,
+  there is no need to indicate which template they correspond to.
+
+*It should be possible to validate templates and template parameters, both values and the schema.*
+
+* Template objects are subject to standard api validation.
+
+*It should also be possible to validate and view the output of the substitution process.*
+
+* The `/processedTemplates` api returns the result of the substitution process, which is itself a Template object that can be validated.
+
+*It should be possible to generate forms for parameterized templates, as discussed in #4210 and #6487.*
+
+* Parameter definitions provide metadata that allows for the construction of form-based UIs to gather parameter values from users.
+
+*It shouldn't be inordinately difficult to evolve templates. Thus, strategies such as versioning and encapsulation should be 
+encouraged, at least by convention.*
+
+* Templates can be versioned via annotations on the template object.
+
+## Alternatives
+
+### Mustache other generic templating language
+
+[Mustache](https://mustache.github.io/) offers a semi-standard generic way to do templating with substitution.  The primary 
+limitation is that the syntax is not valid json (unless all parameters are strings and delimited by quotes), and parameter 
+values must be provided separately.  This means Mustache templates could not be stored and referenced as first class objects.
+
+Other templating systems surveyed had similar limitations:  separated parameters from the template, did not provide generator 
+capabilities, and/or did not use a syntax which could easily be rendered as a Kubernetes API object.
+
+#### Stream transformers
+As defined [here](https://github.com/kubernetes/kubernetes/pull/14993) and [here](https://github.com/kubernetes/kubernetes/pull/14918).
+
+The primary limitation here seems to be that it is a client-side proposal, generators would have to be provided in the client, meaning third party clients/web consoles/etc would have a hard time taking advantage of this mechanism.  It also doesn’t provide a way to create and store new generic templates, you must have a generator defined and then provide appropriate parameters for that generator, so it doesn’t seem like it fills the same space as the templates in this proposal.
+
+Lastly it doesn’t provide a way to list/describe parameters consumed by the template, or provide default values in the template itself, making it hard to build a UI experience around.
+
+#### Deployment Manager
+
+As described [here](https://github.com/kubernetes/deployment-manager)
+
+* Utilizes multiple files with different syntax to define an application topology.
+* Does not define a first class API object.
+* Does not appear to natively offer generation logic for producing parameter values.
+
```

Comment:
DM defines a single template usage format. It's a YAML based list of `resource`, each of which contains a `name`, `type` and `parameters`. True, the template definition can be written either in Jinja, which is simply marked up YAML, or Python, which produces YAML. A Jinja template for a k8s primitive is easily recognizable as a k8s configuration file with mustache markup.

Also, there's no requirement to use both languages in a single example, or to use multiple files, so both assertions are inaccurate.

DM can also take vanilla k8s configuration files as input. And, when a template is expanded, the result is always a flat list of vanilla k8s configuration files, which is readily accessible via the API.

DM doesn't define a k8s API object specifically because a native implementation baked into k8s has been discouraged by the preceding discussion. If that has somehow changed, we could easily make a DM template a native API object. However, we believe there are good reasons not to do that, not least of which is the well established precedent of publishing content in registries, rather than in runtimes. That said, when a configuration is deployed using DM, it is capture and stored in the cluster, providing a record of the deployment, including the parameters and expansion results.

If by native support for generation logic, you mean hard wired, as opposed to programmable, then no, DM doesn't have hard wired generators. We consider hard wired implementations to be an anti-pattern, in general.